sdk/nodejs/runtime/invoke.ts

```diff
@@ -251,28 +251,35 @@ export function call<T>(tok: string, props: Inputs, res?: Resource): Output<T> {
 }
 
 /**
- * Behaves exactly like `call` but returns a Promise instead. Unknowns are not allowed in the response, and an exception
- * will be thrown blaming the provider for violating the contract if the provider returns any unknown values. Secret
- * bits and dependencies returned from the provider are similarly discarded.
+ * Behaves exactly like `call` but returns a Promise instead, to support code generated calls for methods with
+ * XReturnPlainResource. After awaiting the output, extract the resource by field name specified in plainResourceField.
+ * Unknowns are not allowed in the response, and an exception will be thrown blaming the provider for violating the
+ * contract if the provider returns any unknown values. Secret bits and dependencies returned from the provider are
+ * similarly discarded at the moment.
  */
-export function callAsync<T>(tok: string, props: Inputs, res?: Resource): Promise<T> {
+export function callAsync<T>(tok: string, props: Inputs, res: Resource, callAsyncOpts: {
+    plainResourceField: string,
+}): Promise<T> {
     const label = `Calling function: tok=${tok} (callAsync)`;
     log.debug(label + (excessiveDebugOutput ? `, props=${JSON.stringify(props)}` : ``));
 
     return debuggablePromise(new Promise<T>((resolve, reject) => {
-        const resolver = (v: T, _isKnown: boolean, _isSecret: boolean, _deps?: Resource[], err?: Error | undefined) => {
+        const resolver = (v: T, isKnown: boolean, _isSecret: boolean, _deps?: Resource[], err?: Error | undefined) => {
             if (err) {
                 return reject(err);
             }
 
-            // For objects we recur into each field and transform it into an Output, which recursively flattens any
-            // nested Output. This is necessary for type safety, since the expected return types from Call methods
-            // expose properties as prompt values, but the runtime values such as v.foo here are Outputs.
-            //
-            // Unlike call, callAsync does not just transform v into an Output at top level to avoid unkown poisoning,
-            // the situation when any output field being unknown results in the entire result being unknown and
-            // corresponding loss of information. Instead, unknown values are discarded.
-            unsafeUnwrapProperties(v).then(resolve);
+            if (!isKnown) {
+                reject(new Error(`Plain resource method "${tok}" incorrectly returned an unknown Resource value.` +
+                    " This is an error in the provider, please report this to the provider developer."));
+            }
+
+            const extracted = (<any>v)[callAsyncOpts.plainResourceField];
```

Comment:
On a happy path this is already an aws.Provider value in my test. Curious what happens when things go wrong, such as the provider returning an unknown, will try out.